katdal/chunkstore_rados.py

```diff
@@ -99,26 +99,31 @@ def from_config(cls, config, pool, keyring=None, timeout=5.):
             raise StoreUnavailable(str(e))
         return cls(ioctx)
 
-    def get(self, array_name, slices, dtype):
+    def get_chunk(self, array_name, slices, dtype):
         """See the docstring of :meth:`ChunkStore.get`."""
         dtype = np.dtype(dtype)
         key, shape = self.chunk_metadata(array_name, slices, dtype=dtype)
-        num_bytes = int(np.prod(shape)) * dtype.itemsize
+        expected_bytes = int(np.prod(shape)) * dtype.itemsize
         with self._standard_errors(key):
             # Try to read an extra byte to see if data is more than expected
-            data_str = self.ioctx.read(key, num_bytes + 1)
-        if len(data_str) != num_bytes:
+            data_str = self.ioctx.read(key, expected_bytes + 1)
+        actual_bytes = len(data_str)
+        if actual_bytes != expected_bytes:
+            # Get the actual value via stat() to improve error reporting
+            if actual_bytes > expected_bytes:
+                actual_bytes, _ = self.ioctx.stat(key)
             raise BadChunk('Chunk {!r}: dtype {} and shape {} implies an '
                            'object size of {} bytes, got {} bytes instead'
-                           .format(key, dtype, shape, num_bytes, len(data_str)))
+                           .format(key, dtype, shape, expected_bytes,
+                                   actual_bytes))
```

Comment:
Not really related to this PR, but I'm wondering if the `BadChunk` class should store extra fields such as the key, so that given an exception object you can interrogate it to programmatically determine where things went wrong.

Reply:
I like that. It's a little more complicated since `BadChunk` can also happen while constructing the key, but still a useful idea. I've only recently started to store things on the exception, but it is already turning out useful. I think I need a grand error rework in the not-so-near future (once multiple chunk errors are also combined).